README.md
# DelNet

This project is a simple Server-Client system developed in Python using sockets and Pygame. It enables multiple clients to connect to a server and interact with each other by controlling movable squares on a shared canvas. The primary focus of the project is to provide a multiplayer framework, with each player represented as a square that can be moved using the WASD keys. While the core functionality centers around movable squares, additional features can be added to enhance gameplay.

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [Changelog and Roadmap](#changelog-and-roadmap)
- [Contributing](#contributing)
- [License](#license)
- [Acknowledgments](#acknowledgments)

## Installation

You have two options for installing and running DelNet:

### Option 1: Using Pre-built Executables (Recommended)

1. Visit the [Releases](https://github.com/D4N1L0200/DelNet/releases) page on GitHub.
2. Download the latest release, which includes pre-built executables for both the client and server.
3. Run `server.exe` to start the server.
4. Run `client.exe` to start a client.
5. Clients can connect to the server by specifying the server's IP address.

### Option 2: Building from Source

If you prefer to build the project from source code, follow these steps:

1. Clone the repository: `git clone https://github.com/D4N1L0200/DelNet.git`
2. Navigate to the project directory: `cd DelNet`
3. Install dependencies: `pip install pygame`
4. Build the executables? `build.bat`
5. Start the server: `.\dist\server.exe`
6. Run clients: `.\dist\client.exe`

## Usage

- **Starting the Server**: Run `server.exe` to start the server. By default, it will listen on `localhost` (127.0.0.1) on port 8080.
  - If you want the server to be accessible from other devices or through a VPN, type the desired IP address (e.g., the IP address of your VPN connection) when prompted and press Enter.

- **Running Clients**: Execute `client.exe` to start a client. When starting a client, you must specify the IP address of the server you want to connect to interactively in the terminal as follows:
  - When prompted, type the IP address of the server (e.g., the IP address of the server where `server.py` is running) and press Enter.

- **Playing the Game**: Each player is represented by a movable square on the game canvas. Use the WASD keys to move your square around the canvas. The squares of all connected players will be displayed.

## <a id="changelog-and-roadmap"></a> Changelog and Roadmap

- For a detailed list of changes, updates, planned features and bugfixes, please refer to the [Changelog and Roadmap](CHANGELOG-ROADMAP.md).

## Contributing

Contributions to this project are welcome! If you'd like to contribute, please follow these guidelines:

- Check the issue tracker for open issues or create a new one if you identify a problem or have an enhancement idea.
- Fork the repository and create a new branch for your feature or bug fix.
- Ensure your code adheres to Python's PEP 8 style guide.
- Submit a pull request with a clear description of your changes.

## License

This project is licensed under the MIT License. See the [LICENSE](LICENSE) file for details.

## Acknowledgments

Special thanks to the Python community for their contributions to the Python language, and to the Pygame library for enabling the creation of the graphical interface in this project.


CHANGELOG.md
# Changelog

## Version 0.1.0 (2023-09-04)

- First release
- Client experience fully working.

## Version 0.1.1 (2023-09-05)

- Added option to type the server port.
- Now shows the default value for IP and port.
- (BugFix) Added safe checking to the server IP selection.


src/networking.py
import socket
import threading


class Server:
    def __init__(self) -> None:
        self.hostname: str = "127.0.0.1"
        self.port: int = 8080
        self.max_connections: int = 2
        self.socket: socket = None
        self.clients: dict = {}
        print("Server initializing...")

    def init_socket(self, hostname: str, port: str, max_connections: int = 2) -> bool:
        if hostname:
            self.hostname = hostname
        if port:
            try:
                self.port = int(port)
            except ValueError:
                print("Not a valid port.")
                return False
        self.max_connections = max_connections
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.bind((self.hostname, self.port))
        except socket.gaierror:
            print("Not a valid IP.")
            return False
        print(f"Server socket bound on {self.hostname}:{self.port}")
        return True

    def handle_client(self, client_socket: socket, current_client: int) -> None:
        clientid = current_client
        self.clients[clientid] = client_socket
        print(f"New client connected: {client_socket.getpeername()} ({clientid})")

        while True:
            recieved_data: str = client_socket.recv(1024).decode()
            split_data = recieved_data.split("EOM2101")[:-1]
            data: list[tuple] = []
            for item in split_data:
                data.append((str(item[:4]), str(item[4:])))
            for item in data:
                try:
                    match item[0]:
                        case "2400":  # disconnect
                            print(f"Recieved ({clientid}): ", data)  # TODO: REMOVE
                            self.send(clientid, 1400)
                            self.send_others(clientid, 1401, str(clientid))
                            del self.clients[clientid]
                            print(f"Client disconnected: {client_socket.getpeername()}")
                            client_socket.close()
                            return
                        case "2300":  # connect
                            print(f"Recieved ({clientid}): ", data)  # TODO: REMOVE
                            self.send(clientid, 1300, str(list(self.clients.keys()))[1:-1])
                            self.send_others(clientid, 1302, str(clientid))
                        case "2301":  # clientid
                            print(f"Recieved ({clientid}): ", data)  # TODO: REMOVE
                            self.send(clientid, 1301, str(clientid))
                        case "2500":
                            self.send_others(clientid, 1500, item[1])
                        case "2501":
                            self.send_others(clientid, 1501, str(clientid) + ", " + item[1])
                        case _:
                            print(f"Recieved ({clientid}): ", data)
                except Exception as e:
                    print(f"Error handling client: {e}")
                    return

    def start_listening(self) -> None:
        self.socket.listen(self.max_connections)
        print("Server is listening...")
        current_client: int = 1
        while True:
            client_soc, client_address = self.socket.accept()
            client_thread = threading.Thread(target=self.handle_client, args=(client_soc, current_client,))
            client_thread.start()
            current_client += 1

    def send(self, clientid: int, code: int, message: str = "") -> None:
        full_message: str = str(code) + message + "EOM2101"
        self.clients[clientid].send(full_message.encode())
        if code != 1501:
            print(f"Sent ({clientid}): ", code, message)

    def send_others(self, clientid: int, code: int, message: str = "") -> None:
        for otherid, client in self.clients.items():
            if otherid != clientid:
                self.send(otherid, code, message)


class Client:
    def __init__(self) -> None:
        self.hostname: str = "127.0.0.1"
        self.port: int = 8080
        self.socket: socket = None
        self.client_id: int = 0
        self.receive_thread: threading.Thread = threading.Thread()
        self.current_data: list = []
        print("Client initializing...")

    def init_socket(self, hostname: str, port: str) -> bool:
        if hostname:
            self.hostname = hostname
        if port:
            self.port = int(port)
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.socket.connect((self.hostname, self.port))
        except socket.gaierror:
            print("Not a valid IP.")
            return False
        except ConnectionRefusedError:
            print("No server with said IP is online.")
            return False
        except TimeoutError:
            print("Timed out, server took too long to connect.")
            return False
        print(f"Client socket connected to {self.hostname}:{self.port}")
        self.send(2300)
        self.receive_thread = threading.Thread(target=self.receive_data)
        self.receive_thread.start()
        self.send(2301)
        return True

    def receive_data(self) -> None:
        while True:
            try:
                recieved_data: str = self.socket.recv(1024).decode()
                split_data = recieved_data.split("EOM2101")[:-1]
                data: list[tuple] = []
                for item in split_data:
                    data.append((str(item[:4]), str(item[4:])))
                for item in data:
                    match item[0]:
                        case "1300":  # connect + other ids
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            self.current_data.append("1300" + item[1])
                        case "1301":  # client id
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            self.client_id = int(item[1])
                            print("Client ID: ", self.client_id)
                        case "1302":  # other client connected
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            self.current_data.append("1302" + item[1])
                        case "1400":  # disconnect
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            return
                        case "1401":  # other client disconnected
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            self.current_data.append("1401" + item[1])
                        case "1500":  # any game data
                            self.current_data.append("1500" + item[1])
                        case "1501":  # coordenates
                            self.current_data.append("1501" + item[1])
                        case _:
                            print(f"Recieved: ", data)  # TODO: REMOVE
                            pass
            # except ValueError as e:
            #     print(f"Error, a client sent wrong data: {e}")
            except Exception as e:
                print(f"Error: {e}")
                return

    def send(self, code: int, message: str = "") -> None:
        full_message: str = str(code) + message + "EOM2101"
        self.socket.send(full_message.encode())
        if code != 2501:
            print(f"Sent: ", code, message)

    def clear_data(self) -> None:
        self.current_data = []

# when client joins server it needs to know who is already connected


"""
                ==== 1 SERVER ====
1 Info
- 00 Send Connected Clients (SCC)
    - Replies with list of connected clients (name, id) except self
- 01 End Of Message (EOM)
    - Signals the end of a message

2 Error
- 00

3 Connect
- 00 Accept Connection (AC)
    - Replies with all connected client ids
- 01 Send Client ID (SCI)
    - Replies with client id
- 02 Other Client Connected (OCC)
    - Notifies every client that a client has joined and its id

4 Disconnect
- 00 Confirm Disconnection (CD)
    - Confirms the disconnection and stops the thread
- 01 Other Client Disconnected (OCD)
    - Notifies every client that a client has left and its id

5 Game
- 00 Send Echo (SE)
    - Sends previously recieved data to all other clients
- 01 Coordenate Change (CC)
    - Sends coordenates to all other clients and client id


                ==== 2 CLIENT ====
1 Info
- 00 Request Connected Clients (RCC)
    - Requests a list of connected clients (name, id) except self
- 01 End Of Message (EOM)
    - Signals the end of a message

2 Error
- 00

3 Connect
- 00 Request Connection (RC)
    - Requests connection and own client id
- 01 Request ID (RI)
    - Requests own client id

4 Disconnect
- 00 Inform Disconnection (ID)
    - Informs that the client will disconnect shortly

5 Game
- 00 Request Echo (RE)
    - Sends any data to all other clients
- 01 Coordenate Change (CC)
    - Sends coordenates to all other clients and own client id


Transmission Example:
== Server Start ==
2300  C RC
1301. S AC and SCI
2100  C RCC
1101. S SCC
"""

client.py
import pygame
import networking as net


class GameObject:
    def __init__(self, x: int, y: int) -> None:
        self.x = x
        self.y = y


class Square(GameObject):
    def __init__(self, x: int, y: int, size: int, color: tuple) -> None:
        super().__init__(x, y)
        self.size = size
        self.color = color


class Player(Square):
    def __init__(self, x: int, y: int, size: int, speed: int) -> None:
        super().__init__(x, y, size, (255, 255, 255))
        self.speed = speed


class Game:
    def __init__(self) -> None:
        self.width = 800
        self.height = 600
        # self.width = 400
        # self.height = 300

        pygame.init()
        self.screen: pygame.Surface = pygame.display.set_mode((self.width, self.height))
        pygame.display.set_caption("Multiplayer Square")
        self.clock: pygame.time.Clock = pygame.time.Clock()

        self.entities: dict = {"player": None, "players": {}}
        player = Player((self.width - 50) // 2, (self.height - 50) // 2, 50, 5)
        self.entities["player"] = player

    def joinplayer(self, clientid: int) -> None:
        self.entities["players"][clientid] = Square(-100, -100, 50, (255, 0, 0))

    def leaveplayer(self, clientid: int) -> None:
        del self.entities["players"][clientid]


# Initialize Client Socket
HOST: str = input("IP (127.0.0.1): ")
PORT: str = input("Port (8080): ")

client = net.Client()
running = True
if not client.init_socket(HOST, PORT):
    running = False
    input("Press Enter to exit.")
else:
    game = Game()

while running:  # TODO: move to game class
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            client.send(2400)
            running = False

    keys = pygame.key.get_pressed()  # TODO: move all of this to player class
    move_x = (keys[pygame.K_d] - keys[pygame.K_a]) * game.entities["player"].speed
    move_y = (keys[pygame.K_s] - keys[pygame.K_w]) * game.entities["player"].speed

    game.entities["player"].x = max(0, min(game.entities["player"].x + move_x, game.width - game.entities["player"].size))
    game.entities["player"].y = max(0, min(game.entities["player"].y + move_y, game.height - game.entities["player"].size))

    if client.current_data:
        for item in client.current_data:
            match item[:4]:
                case "1300":
                    for cid in map(int, item[4:].split(",")):
                        if cid != client.client_id:
                            game.joinplayer(cid)
                case "1302":
                    game.joinplayer(int(item[4:]))
                case "1401":
                    game.leaveplayer(int(item[4:]))
                case "1500":
                    print("Echo: " + item[4:])
                case "1501":
                    other_id, other_x, other_y = map(int, item[4:].split(","))
                    game.entities["players"][other_id].x = other_x
                    game.entities["players"][other_id].y = other_y
        client.clear_data()

    # if running:
    #     client.send(2500, f"{game.entities[0].x}, {game.entities[0].y}")

    if running:
        client.send(2501, f"{game.entities['player'].x}, {game.entities['player'].y}")

    game.screen.fill((0, 0, 0))
    for cid, square in game.entities["players"].items():
        pygame.draw.rect(game.screen, square.color, (square.x, square.y, square.size, square.size))
    pygame.draw.rect(game.screen, game.entities["player"].color, (game.entities["player"].x, game.entities["player"].y, game.entities["player"].size, game.entities["player"].size))

    pygame.display.flip()
    game.clock.tick(60)


pygame.quit()
